### PR TITLE
Remove root level nested params

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

It makes sure that only Fragments can be used with morph data structures,
updating the types and throwing errors incase the type checks were bypassed

### How to test it?

This is an example on how things should work based on the getstarted app

```ts
strapi.documents('api::restaurant.restaurant').findMany({
      populate: {
        address: {
          populate: {
            cover: {
              fields: '*',
            },
          },
        },
        images: {
          fields: '*',
        },
        dz: {
          on: {
            'default.dish': true,
          },
        },
      },
    });

```
if you try to pass unsupported nested params to Dynamic Zone for example, typescript should show an error.
if typescript was ignored a clear error should be thrown

### CODE REVIEW

**Review by Convly — COMMENTED** [review]
Would it be possible to add unit tests to the convert-query-params? Dow e already have some?

**Comment by Bassel17** [comment]
> Would it be possible to add unit tests to the convert-query-params? Dow e already have some?

yeah, there are some not a lot, there are also some test.todos, so I could attempt to create those test cases

**Comment by Convly** [comment]
> > Would it be possible to add unit tests to the convert-query-params? Dow e already have some?
> 
> yeah, there are some not a lot, there are also some test.todos, so I could attempt to create those test cases

Would be awesome!

**Review by Convly — CHANGES_REQUESTED** [review]
You'll also need to remove the support for the deprecated syntax in the query-populate DZ handler.
`https://github.com/strapi/strapi/blob/v5/main/packages/core/utils/src/traverse/query-populate.ts`
```ts
// Handle legacy DZ params
        let newProperties: unknown = omit('on', value);

        for (const componentUID of components) {
          const componentSchema = getModel(componentUID);

          const properties = await recurse(
            visitor,
            { schema: componentSchema, path, getModel },
            value
          );
          newProperties = merge(newProperties, properties);
        }

        Object.assign(newValue, newProperties);
```

File: packages/core/utils/src/convert-query-params.ts [hunk 1]
```diff
@@ -343,6 +343,15 @@ const createTransformer = ({ getModel }: TransformerOptions) => {
...
     return Object.entries(populate).reduce((acc, [key, subPopulate]) => {
+      if (_.isString(subPopulate)) {
+        try {
+          const subPopulateAsBoolean = parseType({ type: 'boolean', value: subPopulate });
+          return { ...acc, [key]: subPopulateAsBoolean };
+        } catch {
+          // ignore
+        }
+      }
+
```
Comment (Convly) [inline packages/core/utils/src/convert-query-params.ts:354]: Maybe it's adding a comment explaining why we need this new parse op. (aka "true" can be passed as a string)

File: packages/core/utils/src/convert-query-params.ts [hunk 2]
```diff
@@ -357,42 +366,29 @@ const createTransformer = ({ getModel }: TransformerOptions) => {
...
-        const populates = attribute.components
-          .map((uid) => getModel(uid))
-          .map((schema) => convertNestedPopulate(subPopulate, schema))
-          .map((populate) => (populate === true ? {} : populate)) // cast boolean to empty object to avoid merging issues
-          .filter((populate) => populate !== false);
-
-        if (isEmpty(populates)) {
-          return acc;
+      if (isAllowedAttributeForFragmentPopulate) {
+        if (hasFragmentPopulateDefined(subPopulate)) {
```
Comment (Convly) [inline packages/core/utils/src/convert-query-params.ts:370]: Currently the hasFragmentPopulateDefined function only verifies that an on property is defined and non nil, but it doesn't make sure no other props are present. Might be worth adding a check to the function directly?

Comment (Bassel17) [reply packages/core/utils/src/convert-query-params.ts:370]: not sure I understood, you mean it doesn't check if it has any properties ?

Comment (Convly) [reply packages/core/utils/src/convert-query-params.ts:370]: Basically, we need to check that only `on` is defined in the sub populate, no other key should be accepted

**Review by innerdvations — COMMENTED** [review]
Reviewed and nothing to add on the code that hasn't already been said.

There's a linting error, and it looks like some API tests need to be updated maybe because they were using the deprecated format.

**Review by Convly — CHANGES_REQUESTED** [review]
Seems like we still have an issue with populate: '*'

https://github.com/strapi/strapi/actions/runs/9189884447/job/25273258553?pr=20172

File: tests/api/core/strapi/api/populate/sanitize.test.api.js [hunk 3]
```diff
@@ -143,9 +145,13 @@ describe('Sanitize populated entries', () => {
...
       const { status, body } = await contentAPIRequest.get(`/upload/files/${file.id}`, {
-        qs: { populate: { related: { populate: '*' } } },
+        qs: {
+          populate: {
+            related: true,
+          },
+        },
       });
-
+      console.log(body);
```
Comment (Convly) [inline tests/api/core/strapi/api/populate/sanitize.test.api.js:154]:
```suggestion
```

**Review by Convly — APPROVED** [review]

**Pull request merged**